### ISSUE TITLE
🚀 Implement Backward Pass for All Autograd Operations

### ISSUE THREAD

**Issue by sakshambedi** [sakshambedi/PyGrad#11]
# 🚀 Implement Backward Pass for All Autograd Operations

## 📋 Description

Currently, the autograd system has forward passes implemented for all operations (Add, Sub, Mul, Div, Pow, Neg), but the backward passes are incomplete. This is critical for enabling gradient computation and automatic differentiation.

## 🎯 Acceptance Criteria

- [x] Implement `backward()` method for `Sub` operation
- [x] Implement `backward()` method for `Mul` operation  
- [ ] Implement `backward()` method for `Div` operation
- [ ] Implement `backward()` method for `Pow` operation
- [ ] Implement `backward()` method for `Neg` operation
- [ ] Add comprehensive tests for gradient computation
- [ ] Ensure gradient flows correctly through computational graph
- [ ] Add edge case handling (e.g., division by zero in gradients)

### Current State
- Forward passes: ✅ Complete
- Backward passes: ❌ Missing implementations (only Add is implemented)
- Context saving: ✅ Implemented for Add and Neg

### Implementation Requirements

**Subtraction (Sub)**
```python
# dL/da = grad_output, dL/db = -grad_output
```

**Multiplication (Mul)**
```python
# dL/da = b * grad_output, dL/db = a * grad_output
# Need to save a, b in context
```

**Division (Div)**
```python
# dL/da = grad_output / b, dL/db = -a * grad_output / (b^2)
# Need to save a, b in context
```

**Power (Pow)**
```python
# dL/da = b * (a^(b-1)) * grad_output
# dL/db = (a^b) * ln(a) * grad_output
# Need to save a, b in context
```

**Negation (Neg)**
```python
# dL/da = -grad_output
```

## 🧪 Testing Requirements

- [ ] Test gradients match numerical gradients (finite differences)
- [ ] Test gradient flow through mixed operations
- [ ] Test edge cases (zeros, ones, negative numbers)
- [ ] Test chain rule application
- [ ] Performance benchmarks for backward pass

## 📁 Files to Modify

- `grad/autograd/ops.py` - Main implementation
- `tests/test_autograd.py` - New test file needed
- `tests/gradient_check.py` - Numerical gradient verification

## 🔗 Related Issues

- Depends on: Buffer management system (completed)
- Blocks: Gradient accumulation (#03)
- Blocks: Neural network layers (#08)

## 🏷️ Labels

- `enhancement`
- `autograd`
- `high-priority`
- `core-functionality`

## 📈 Priority

**High** - Critical for autograd functionality
